entry/src/entry.rs

```diff
@@ -339,6 +396,155 @@ pub fn verify_transactions(
+            }
+            Err(err) => {
+                return Err(err);
+            }
+        }
+    }
+
+    let verify_func = {
+        move |versioned_tx: VersionedTransaction| -> Result<SanitizedTransaction> {
+            verify(
+                versioned_tx,
+                TransactionVerificationMode::HashAndVerifyPrecompiles,
+            )
+        }
+    };
+    let entries = verify_transactions(entries, Arc::new(verify_func));
+    match entries {
+        Ok(entries) => {
+            let num_transactions: usize = entries
+                .iter()
+                .map(|entry: &EntryType| -> usize {
+                    match entry {
+                        EntryType::Transactions(transactions) => transactions.len(),
+                        EntryType::Tick(_) => 0,
+                    }
+                })
+                .sum();
+
+            if num_transactions == 0 {
+                return Ok(EntrySigVerificationState {
+                    verification_status: EntryVerificationStatus::Success,
+                    entries: Some(entries),
+                    device_verification_data: DeviceSigVerificationData::Cpu(),
+                    gpu_verify_duration_us: 0,
+                });
+            }
+            let mut packets = Packets::new_with_recycler(
+                verify_recyclers.packet_recycler,
+                num_transactions,
+                "entry-sig-verify",
+            );
+            unsafe {
+                packets.packets.set_len(num_transactions);
+            }
+
+            {
+                let entry_txs: Vec<&SanitizedTransaction> = entries
+                    .iter()
+                    .filter_map(|entry_type| match entry_type {
+                        EntryType::Tick(_) => None,
+                        EntryType::Transactions(transactions) => Some(transactions),
+                    })
+                    .flatten()
+                    .collect::<Vec<_>>();
+
+                let entry_tx_iter = entry_txs
+                    .into_par_iter()
+                    .map(|tx| tx.to_versioned_transaction());
+
+                let res = packets
+                    .packets
+                    .par_iter_mut()
+                    .zip(entry_tx_iter)
+                    .all(|pair| {
+                        pair.0.meta = Meta::default();
+                        Packet::populate_packet(pair.0, None, &pair.1).is_ok()
+                    });
+
+                if !res {
+                    return Err(TransactionError::SignatureFailure);
+                }
+            }
+
+            let mut packets = vec![packets];
+            let tx_offset_recycler = verify_recyclers.tx_offset_recycler;
+            let out_recycler = verify_recyclers.out_recycler;
+            let gpu_verify_thread = thread::spawn(move || {
+                let mut verify_time = Measure::start("sigverify");
+                sigverify::ed25519_verify(&mut packets, &tx_offset_recycler, &out_recycler, false);
```

Comment:
What if we use this code path for both CPU and GPU? `sigverify::ed25519_verify` already checks for perf apis and that the batch size is not too low before using the GPU

Reply:
Feel free to punt this to another PR

Reply:
I'd prefer not to since, there's non-trivial CPU and memory overhead needed to get the transactions into packets (this is why the CPU-to-GPU crossover point in this function is higher than in sigverify::ed25519_verify itself), and if sigverify::ed25519_verify does it in the CPU, the batch size (<64 transactions) would be too small for the async verification to be of benefit, and the extra overhead from getting the transactions ready for sigverify::ed25519_verify would still apply.

Reply:
In replay stage we call [`Bank::verify_transaction`](https://github.com/solana-labs/solana/blob/de8edad30b3ca92a6936f277093033cee4eb6214/runtime/src/bank.rs#L5502) which in turn will serialize the transaction message and result in some memory overhead. If we switch to lazy sigverify, we could also do lazy hash, and avoid serializing the transaction message there. We would instead always serialize transactions in a memory efficient way with a packet recycler.

Reply:
Ok, sounds good but punting to a separate PR.